book-content/chapters/16-the-utils-folder.md

```diff
@@ -0,0 +1,1527 @@
+
+Let's put what we've learned into practice.
+
+Consider this function called `getFirstElement` that takes an array and returns the first element:
+
+```typescript
+const getFirstElement = (arr: any[]) => {
+  return arr[0];
+};
+```
+
+This function is dangerous. Because it takes an array of `any`, it means that the thing we get back from `getFirstElement` is also `any`:
+
+```ts twoslash
+const getFirstElement = (arr: any[]) => {
+  return arr[0];
+};
+
+// ---cut---
+const first = getFirstElement([1, 2, 3]);
+//    ^?
+```
+
+As we've seen, `any` can cause havoc in your code. Anyone who uses this function will be unwittingly opting out of TypeScript's type safety. So, how can we fix this?
+
+We need TypeScript to understand the type of the array we're passing in, and use it to type what's returned. We need to make `getFirstElement` generic:
+
+To do this, we'll add a type parameter `TMember` before the function's parameter list, then use `TMember[]` as the type for the array:
+
+```typescript
+const getFirstElement = <TMember>(arr: TMember[]) => {
+  return arr[0];
+};
+```
+
+Just like generic functions, it's common to prefix your type parameters with `T` to differentiate them from normal types.
+
+Now when we call `getFirstElement`, TypeScript will infer the type of `` based on the argument we pass in:
+
+```ts twoslash
+const getFirstElement = <TMember>(arr: TMember[]) => {
+  return arr[0];
+};
+// ---cut---
+const firstNumber = getFirstElement([1, 2, 3]);
+//    ^?
+const firstString = getFirstElement(["a", "b", "c"]);
+//    ^?
+```
+
+Now, we've made `getFirstElement` type-safe. The type of the array we pass in is the type of the thing we get back.
+
+### Debugging The Inferred Type Of Generic Functions
+
+When you're working with generic functions, it can be hard to know what type TypeScript has inferred. However, with a carefully-placed hover, you can find out.
+
+When we call the `getFirstElement` function, we can hover over the function name to see what TypeScript has inferred:
+
+```ts twoslash
+const getFirstElement = <TMember>(arr: TMember[]) => {
+  return arr[0];
+};
+// ---cut---
+const first = getFirstElement([1, 2, 3]);
+//            ^?
+```
+
+We can see that within the angle brackets, TypeScript has inferred that `TMember` is `number`, because we passed in an array of numbers.
+
+This can be useful when you have more complex functions with multiple type parameters to debug. I often find myself creating temporary function calls in the same file to see what TypeScript has inferred.
+
+### Type Parameter Defaults
+
+Just like generic types, you can set default values for type parameters in generic functions. This can be useful when runtime arguments to the function are optional:
+
+```typescript
+const createSet = <T = string>(arr?: T[]) => {
+  return new Set(arr);
+};
+```
+
+Here, we set the default type of `T` to `string`. This means that if we don't pass in a type argument, TypeScript will assume `T` is `string`:
```

Comment:
only if it fails to infer anything - the sentence here sounds like a rule that could be applied more broadly